---
title: "DEGs with RUV normalization with FAKE condition - 2 step analysis"
author: "apietrelli"
output: 
  html_document:
    fig_height: 8
    fig_width: 15
    theme: united
    toc: yes
---

DESCRIPTION

# Date of anlaysis and analysis filename

```{r timestamp}
date()
filename="DESeq2_2Step_normalization_RUV.Rmd"
print(filename)
outdir = "RUV_Normalization_27042016/"
```

---

# Step1 DEGs considering "FAKE" condition as Proliferation 

## Preparing input files

Loading library

```{r loading_lib, message=FALSE}
library("BiocParallel")
register(MulticoreParam(10))
library('DESeq2')
library('knitr')
library("ggplot2")
```

## Set Filename and directory

9 samples will be analyzed for differential expression analysis with FAKE condition

```{r loading_files}
csv_path='SampleSheet_HSC_stranded_Proliferation.tsv'
ref_condition="A"

### Example sample sheet
# Sample sheet is SPACE-DELIMITED, with header
# Example
# sampleName fileName condition1
# HET_721 HET_721.HQ.counts HET
# HET_729 HET_729.HQ.counts HET
sampleTable <- read.table(csv_path,header=TRUE)
sample_dir='../mapping/'

```

## Read files and generate "dds"

- The fake condition was determined by analyzing the PCA plot with all the samples.

- The condition are "A" and "B" -> for High- and Low- proliferative samples

```{r deseq, message=FALSE}
# DESIGN
# MULTI FACTOR -> Condition + proliferation status
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,directory = sample_dir, design= ~ condition + type)

# DESeq worker
dds <- DESeq(ddsHTSeq, parallel=TRUE)
# All samples > 1 in gene count
dds<-dds[rowSums(counts(dds)) > 1,]

# Extracting counts
raw.counts <- as.data.frame(counts( dds ))
normalized.counts <- as.data.frame(counts( dds, normalized=TRUE ))

# Relevel the condition
dds$type <- relevel(dds$type, ref=ref_condition)
res <- results(dds, parallel=TRUE)

# Stabilizing the variance (takes long...)
rld <- rlog(dds)
```

## PCA Plot 
See the clusterization of the samples by "type" and WT- HET-condition

```{r pca_fake, message = F}
data_PCA <- plotPCA(rld, intgroup = c("type","condition"), returnData=TRUE)
percentVar <- round(100 * attr(data_PCA, "percentVar"))
library("ggplot2")
qplot(PC1, PC2, color=condition, data=data_PCA, label=rownames(data_PCA)) +
  #scale_colour_gradient(low="black", high="red") +
  geom_point(aes (shape = factor(type) ), size=4) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + geom_text(vjust = 1.7, nudge_x = 0.05, check_overlap = T)

```

# Extraction DEGs from B vs A condition

Gene vector for RUV normalization

```{r pca, message = F}
# First 5000 genes ordered by padj
empirical=rownames(as.data.frame(res[order(res$padj),]))[1:5000]
```

# Normalization step with RUV seq

## Extract raw counts for "real" condition samples

```{r load_reaSamples, message = F}
csv_path='SampleSheet_HSC_outlierOut.tsv'
ref_condition="WT"
outdir="RUV_Normalization_27042016"

sampleTable <- read.table(csv_path,header=TRUE)
sample_dir='../mapping/'
```

```{r extractRaw_realsamples, message = F}
## DESIGN
# Single factor -> CONDITION
#ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,directory = sample_dir, design= ~ condition)
# MULTI FACTOR -> Condition + proliferation status expressed in LogCellCount
#ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,directory = sample_dir, design= ~ LogCellCount + condition)
# MULTI FACTOR 2 -> Condition + proliferation status expressed in million cell (MCellCount)
ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,directory = sample_dir, design= ~ MCellCount + condition)

# DESeq worker
dds <- DESeq(ddsHTSeq, parallel=TRUE)
## Pre-filtering
# All samples > 1 in gene count
dds<-dds[rowSums(counts(dds)) > 1,] 

# Extracting counts
raw.counts <- as.data.frame(counts( dds ))
```

## Normalization with RUV

```{r normalization, message = F}
library(RUVSeq)
# Spikes to use as NORMALIZATORS
spikes = empirical
# Levels for condition
x = colData(dds)@listData$condition
# Extracting counts as "set" for RUV seq
set <- newSeqExpressionSet(as.matrix(raw.counts),
                           phenoData = data.frame(x, row.names=colnames(raw.counts)))

set_RUVg <- RUVg(x = set, spikes, k = 1)

plotRLE(set_RUVg, outline=FALSE, ylim=c(-4, 4))
plotPCA(set_RUVg, cex=1.2)
normalized.counts.RUV = set_RUVg@assayData$normalizedCounts

```

# DESeq with normalized RUV counts

## DESeq with normalized counts performed by RUVSeq

```{r DESeq_realSamples, message = F}

condition <- factor(rep(c("HET","WT"), each=4 ))
ddsRUV = DESeqDataSetFromMatrix(normalized.counts.RUV, DataFrame(condition), ~ condition)

# DESeq worker
dds <- DESeq(ddsRUV, parallel=TRUE)
## Pre-filtering
# All samples > 1 in gene count
dds<-dds[rowSums(counts(dds)) > 1,] 

# Extracting counts
raw.counts <- as.data.frame(counts( dds ))
normalized.counts <- as.data.frame(counts( dds, normalized=TRUE ))

# Relevel the condition
dds$condition <- relevel(dds$condition, ref=ref_condition)
res <- results(dds, parallel=TRUE)
head(res)
# Summary of DEGs
summary(res)

# Stabilizing the variance (takes long...)
rld <- rlog(dds)

# Saving rld as 
saveRDS(object = rld, paste(outdir, "HSC_RUV_rld.rds", sep="/"))
```

## Generate "means" DF according to treatment (condition WT/HET)

```{r means_df, message=F}
# Variable for the condition
populations = levels(dds$condition)

#Generate 'means' dataframe
len = length(rowMeans(counts(dds,normalized=TRUE)[,dds$condition == populations[19]]))
means = data.frame(matrix(NA, nrow = len))
count = 0
for (pop in populations){
  print(pop)
  tryCatch({means[pop] = rowMeans(counts(dds,normalized=TRUE)[,dds$condition == pop])},
           error = function(e) {
             means[pop] = counts(dds,normalized=TRUE)[,dds$condition == pop]
           })    
}
# Remove NA initial column from 'means'
means = means[,2:length(means)]
rownames(means) = rownames(normalized.counts)

means$ensembl_gene_id = rownames(means)
```

## Gene annotation Mouse

```{r annotation, message=F}
#retrieve gene data
library("biomaRt")
gene_set = rownames(means)
ensembl=useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset = "mmusculus_gene_ensembl",
                host = "feb2014.archive.ensembl.org",
                path = "/biomart/martservice")
genes_information = getBM(c("mgi_symbol",
                            "entrezgene",
                            "uniprot_genename",
                            "description",
                            "gene_biotype",
                            "chromosome_name",
                            "strand",
                            "start_position",
                            "end_position",
                            "ensembl_gene_id"),
                          "ensembl_gene_id",
                          gene_set, ensembl)
genes_with_information = merge(x = means, y = genes_information, by = "ensembl_gene_id", all.x = TRUE)
```

## Generate files for GSEA

```{r export_for_GSEA}
gene_set = rownames(normalized.counts)
res_information = getBM(c("ensembl_gene_id",
                          "mgi_symbol"),
                          "ensembl_gene_id",
                          gene_set, ensembl)
tmpfile = cbind(ensembl_gene_id = rownames(normalized.counts), normalized.counts)
rownames(tmpfile) <- NULL
gseafilename = merge(res_information, tmpfile, by = "ensembl_gene_id")[,-1]
gseafilename = gseafilename[gseafilename$mgi_symbol != "",]

write.table(gseafilename, paste(outdir, "HSC_Results_GSEA_input.txt", sep = "/"), sep="\t", quote=F, row.names=F)

## Pre-Rank file based on log2FC
res_df = as.data.frame(res)
res_df = cbind(ensembl_gene_id = rownames(res_df), res_df)
rownames(res_df) <- NULL
rankGSEA = merge(res_information, res_df, by = "ensembl_gene_id")[,-1]
rankGSEA = rankGSEA[rankGSEA$mgi_symbol != "",]
rankGSEA = rankGSEA[order(rankGSEA$log2FoldChange, decreasing = F),]

write.table(rankGSEA[,c("mgi_symbol","log2FoldChange")],
            paste(outdir, "HSC_Results_GSEA_input.rnk", sep = "/"),
            sep="\t",
            quote=F,
            row.names=F)
```

## Select significant genes

```{r sigGenes}
alpha=0.1
log2FC_th=0.58
resSig <- res[which(res$padj <=alpha & (res$log2FoldChange>=log2FC_th | res$log2FoldChange<=-log2FC_th)),]
resSig$ensembl_gene_id=rownames(as.data.frame(resSig))
downDEGs <- resSig[which(resSig$log2FoldChange < 0),]
upDEGs <- resSig[which(resSig$log2FoldChange > 0),]
```

# Write the results

## Write DEGs results with annotation

```{r DEGs_results}
# Significant gene annotation
resSig_annot = merge(as.data.frame(resSig), genes_with_information, by="ensembl_gene_id")
write.table(resSig_annot, paste(outdir, "HSC_DEGs_annotation.xls", sep = "/"),sep="\t", quote=F, row.names=F)

# UP genes
write.table(resSig_annot[which(resSig_annot$log2FoldChange>=0),],
            paste(outdir, "HSC_DEGs_UP_annotation.xls", sep = "/"),
            sep="\t", quote=F, row.names=F)
# DOWN genes
write.table(resSig_annot[which(resSig_annot$log2FoldChange<=0),], paste(outdir, "HSC_DEGs_DOWN_annotation.xls", sep = "/"),
            sep="\t", quote=F, row.names=F)
```

## Write ALL genes with DESeq statistics (Useful for GSEA)

```{r ALL_results}
# All genes with annotation
gene_set = rownames(res)
res_information = getBM(c("ensembl_gene_id",
                          "mgi_symbol"),
                          "ensembl_gene_id",
                          gene_set, ensembl)
res2write = cbind(ensembl_gene_id = rownames(res), as.data.frame(res))
res2write = merge(res2write, res_information, by = "ensembl_gene_id")
write.table(res2write, paste(outdir, "HSC_DESeq_Results.xls", sep = "/"), sep="\t", quote=F, row.names=F)
```

```{r DEGs_for_GSEA}
rankDEG_GSEA = resSig_annot[resSig_annot$mgi_symbol != "" & !is.na(resSig_annot$mgi_symbol),]
rankDEG_GSEA = rankDEG_GSEA[order(rankDEG_GSEA$log2FoldChange, decreasing = F),]

write.table(rankDEG_GSEA[,c("mgi_symbol","log2FoldChange")],
            paste(outdir, "HSC_DEGs_GSEA_input.rnk", sep = "/"),
            sep="\t",
            quote=F,
            row.names=F)

```



# Visualization

## Preparing data for plot

```{r preparing_data_for_plot, echo=F}
# Which genes are significant
ens_genesSig=rownames(resSig)

# Extract annotation
geneSig = genes_with_information[genes_with_information$ensembl_gene_id %in% ens_genesSig,]

# Create DF for means with geneSig annotation
means$significant[means$ensembl_gene_id %in% ens_genesSig] <- 1
means$significant[!(means$ensembl_gene_id %in% ens_genesSig)] <- 0
genes_with_information$significant[genes_with_information$ensembl_gene_id %in% ens_genesSig] <- 1
genes_with_information$significant[!(genes_with_information$ensembl_gene_id %in% ens_genesSig)] <- 0
```


## Plot counts of specific genes
- "TNF","IL1B","IL6","CXCL1","CXCL10","INSR","LOXL2","CCL2","COL1A1","ACTA2","PECAM1","GAPDH","ACTB", "PDGFB", "PDGFRB"

* IFNG - ENSMUSG00000055170
-> 0 counts in all the samples

```{r plotcounts_specific, message = F}
# Plot contus of specific genes
library(ggplot2)
### List of specific genes
## From Marica presentation
gene_list = c("TNF","IL1B","IL6","CXCL1","CXCL10","INSR","LOXL2","CCL2","COL1A1","ACTA2","PECAM1","GAPDH","ACTB", "PDGFB", "PDGFRB")
## IFNG not work
## ENSMUSG00000055170
## 0 counts in all the samples

ensgene_list = sapply(gene_list, function(x){
  unique(genes_with_information[which(toupper(genes_with_information$mgi_symbol)==x),]$ensembl_gene_id)
  })
# Plot gene counts
#par(mfrow=c(3,1),mar=c(3,5,3,1))
plots = lapply(names(ensgene_list),function(x){
  if(length(ensgene_list[[x]]) != 0){
    data <- plotCounts(dds, gene=ensgene_list[[x]], intgroup=c("condition"), normalized=T, returnData=TRUE)
    ggplot(data, aes(x=condition, y=count, colour = condition)) +
      scale_y_log10() + 
      #scale_colour_continuous(high= "red", low =  "black") +
      geom_point(position=position_jitter(width=.1,height=0), size=5)+
      geom_text(aes(label=rownames(data)),
                #position = position_jitter(width=.5,height=0)+
                hjust=2, 
                vjust=0)+
      ggtitle(paste(as.character(x),as.character(ensgene_list[[x]]), sep = " - "))
    }
})
data <- plotCounts(dds, gene=ensgene_list[[1]], intgroup=c("condition"), normalized=T, returnData=TRUE)
plots
# Gene list are significant?
gene_list_res=as.data.frame(res[rownames(res) %in% unique(genes_with_information[which(toupper(genes_with_information$mgi_symbol) %in% gene_list ),]$ensembl_gene_id),])
gene_list_res=cbind(ensembl_gene_id=rownames(gene_list_res), gene_list_res)
gene_list_res = merge(x = res_information, y = gene_list_res, by = "ensembl_gene_id")
gene_list_res$significant=gene_list_res$padj<alpha
kable(gene_list_res)
```

## HSC markers plot

```{r plotcounts_HSC_marker, message = F}
# Plot contus of specific genes
library(ggplot2)
### List of specific genes
### http://www.nature.com/nprot/journal/v10/n2/pdf/nprot.2015.017.pdf
### HIGH-retinoid --> GOOD HSC purification --> UP regulation of markers
### LOW-retinoid --> BAD HSC purification --> DOWN regulation of markers
### Activation marker genes
gene_list = c("LRAT","LHX2","HAND2","PDGFRB","VIM", "GFAP", "DES", "PPARG","TGFBR1","ACTA2","BAMBI")

ensgene_list = sapply(gene_list, function(x){
  unique(genes_with_information[which(toupper(genes_with_information$mgi_symbol)==x),]$ensembl_gene_id)
  })
# Plot gene counts
#par(mfrow=c(3,1),mar=c(3,5,3,1))
plots = lapply(names(ensgene_list),function(x){
  if(length(ensgene_list[[x]]) != 0){
    data <- plotCounts(dds, gene=ensgene_list[[x]], intgroup=c("condition"), normalized=T, returnData=TRUE)
    ggplot(data, aes(x=condition, y=count, colour = condition)) +
      scale_y_log10() + 
      #scale_colour_continuous(high= "red", low =  "black") +
      geom_point(position=position_jitter(width=.1,height=0), size=5)+
      geom_text(aes(label=rownames(data)),
                #position = position_jitter(width=.5,height=0)+
                hjust=2, 
                vjust=0)+
      ggtitle(paste(as.character(x),as.character(ensgene_list[[x]]), sep = " - "))
    }
})
data <- plotCounts(dds, gene=ensgene_list[[1]], intgroup=c("condition"), normalized=T, returnData=TRUE)
plots
# Gene list are significant?
gene_list_res=as.data.frame(res[rownames(res) %in% unique(genes_with_information[which(toupper(genes_with_information$mgi_symbol) %in% gene_list ),]$ensembl_gene_id),])
gene_list_res=cbind(ensembl_gene_id=rownames(gene_list_res), gene_list_res)
gene_list_res = merge(x = res_information, y = gene_list_res, by = "ensembl_gene_id")
gene_list_res$significant=gene_list_res$padj<alpha
kable(gene_list_res)
```

## PCA Real Samples

```{r pca_real, message = F}
data_PCA <- plotPCA(rld, intgroup = c("condition"), returnData=TRUE)
percentVar <- round(100 * attr(data_PCA, "percentVar"))
library("ggplot2")
qplot(PC1, PC2, color=condition, data=data_PCA, label=rownames(data_PCA)) +
  #scale_colour_gradient(low="black", high="red") +
  #geom_point(aes (size=MCellCount*5) )+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + geom_text(vjust = 1.7, nudge_x = 0.05, check_overlap = T)

```

## Scatter plot

```{r scatter, message = F}
library(ggplot2)
colScale <- scale_colour_manual(name = "significant", values = c("gray","red"))
ggplot(genes_with_information, aes(WT, HET, colour=as.factor(significant), alpha = as.factor(significant))) +
  geom_point() +
  colScale +
  # label
  #geom_text(aes(label=ifelse((significant==1),toupper(as.character(mgi_symbol)),'')),
  #          hjust=0, 
  #          vjust=0, 
  #          check_overlap = T) +
  scale_y_log10() +
  scale_x_log10()

```

## Gene clustering

```{r gene_clust, message = F}
# DEGs Gene
library("genefilter")
library("pheatmap")
alpha=0.1
resSig <- res[which(res$padj <=alpha),]

mat <- assay(rld)[ which(rownames(rld) %in% rownames(resSig)),]
mat <- mat - rowMeans(mat)
# Create Gene name dataframe
#df_gene <- as.data.frame(rownames(mat))
#colnames(df_gene) <- "ensembl_gene_id"
#df_gene = merge(x = df_gene, y = genes_information, by = "ensembl_gene_id", all.x = TRUE)
# fix NA of BLANK values with ENSID
#idx_to_fix = which(df_gene$mgi_symbol == "" | is.na(df_gene$mgi_symbol))
#df_gene[idx_to_fix,]$mgi_symbol = as.vector(df_gene[idx_to_fix,]$ensembl_gene_id)
#rownames(mat) = unique(df_gene[match(df_gene$ensembl_gene_id,rownames(mat)),]$mgi_symbol)
# Create condition dataframe
df <- as.data.frame(colData(rld)[,"condition"])
colnames(df) <- "condition"
rownames(df) <- rownames(colData(rld))
# Plot the heatmap
pheatmap(mat, annotation_col = df, show_rownames = F)
```

## Extract the gene list relative to the tree cut

```{r extract_gene_clust, message = F}
library(gplots)

# Calculate hclust from distance matrix (mat)
hr = hclust(dist(mat))
# Get the tree cut - Gene belonging to treecut 
mycl=cutree(hr,3)

# choose a color palette for the heat map
myheatcol <- rev(redgreen(75))

# get a color palette equal to the number of clusters
clusterCols <- rainbow(length(unique(mycl)))

# create vector of colors for side bar
myClusterSideBar <- clusterCols[mycl]

# draw the heat map with heatmap.2
heatmap.2(mat, main="Hierarchical Cluster on DEGs", Rowv=as.dendrogram(hr), Colv=NA, dendrogram="row", scale="row", col=myheatcol, density.info="none", trace="none", RowSideColors= myClusterSideBar, margins = c(7,4), labRow = F)

# examine the cluster membership by it's order
# in the heatmap
#mycl[hr$order]
```


```{r single_gene_expression}
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols))
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

boxplot_counts <- function(GeneName, res_information, dds){
  g=res_information[res_information$mgi_symbol==GeneName,]$ensembl_gene_id
  data <- plotCounts(dds, gene=g, intgroup=c("condition"), normalized=T, returnData = T)
  plot = ggplot(data, aes(x=condition, y=count/1000)) +
    #scale_y_log10(
    #  breaks = scales::trans_breaks("log10", function(x) 10^x),
    #  labels = scales::trans_format("log10", scales::math_format(10^.x))
    #) + 
    #annotation_logticks(sides = "l") +
    theme_bw() +
    #scale_colour_continuous(high= "red", low =  "black") +
    #geom_point(position=position_jitter(width=.1,height=0), size=5)+
    #geom_text(aes(label=rownames(data)),
              #position = position_jitter(width=.5,height=0)+
    #          hjust=2, 
    #          vjust=0)+
    geom_boxplot(width = 0.7) +
    xlab("") +
    ylab("Gene expression\n(Normalized read count x 1000)") +
    ggtitle(as.character(GeneName))
  return(plot)
}

## LOXL2
loxl2_boxplot = boxplot_counts("Loxl2", res_information, dds)
## COL4A6
col4a6_boxplot = boxplot_counts("Col4a6", res_information, dds)

pdf("Plot/LOXL2_COL4A6.boxplot.pdf", paper ="a4")
multiplot(loxl2_boxplot, col4a6_boxplot)
dev.off()

# Export counts
for (gene in c("Loxl2","Col1a1","Col4a6","Col4a1")){
  g=res_information[res_information$mgi_symbol==gene,]$ensembl_gene_id
  data <- plotCounts(dds, gene=g, intgroup=c("condition"), normalized=T, returnData = T)
  write.csv2(plotCounts(dds, gene=g, intgroup=c("condition"), normalized=T, returnData = T), 
              paste("RUV_Normalization_27042016/",gene,"_counts.xls",sep=""), quote=F, row.names = T)
}
# Export pvalues
write.table(res2write[res2write$mgi_symbol %in% c("Loxl2","Col1a1","Col4a6","Col4a1"),],"RUV_Normalization_27042016/GeneList_info.xls", quote=F,row.names = F,sep="\t")
```

